src/http_server.c

```diff
@@ -0,0 +1,150 @@
+#include <netinet/in.h>
+#include <string.h>
+#include <unistd.h>
+#include <ctype.h>
+#include <netinet/tcp.h>
+
+#include "config.h"
+#include "log.h"
+#include "tunnel.h"
+
+// Convenience function to send a response with a user readable status message and the
+// request URL shown for debug purposes.
+// The response is shown in the user's browser after being redirected from the Fortinet Server.
+static void send_status_response(int socket, const char *userMessage) {
+    const char *reply = "HTTP/1.1 200 OK\r\n"
+            "Content-Type: text/plain\r\n"
+            "Content-Length: %lu\r\n"
+            "Connection: close\r\n\r\n"
+            "%s"; // User readable message
+
+    int requiredSize = snprintf(NULL, 0, reply, strlen(userMessage), userMessage) + 1;
+    char *buffer = alloca(requiredSize);
+    snprintf(buffer, requiredSize, reply, strlen(userMessage), userMessage);
+
+    ssize_t write_result = write(socket, buffer, strlen(buffer));
+    (void)write_result;
+}
+
+int process_request(int new_socket, char *id) {
+    log_info("Processing HTTP SAML request\n");
+
+    int flag = 1;
+    setsockopt(new_socket, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int));
+
+    // Read the request
+    char request[1024];
+    ssize_t read_result = read(new_socket, request, sizeof(request));
+
+    // Check for '=id' in the response
+    // If the recevied request from the server is larger than the buffer, the result will not be null-terminated.
+    // Causing strlen to behave wrong.
+    if (read_result < 0 || read_result == sizeof(request)) {
+        log_error("Bad request\n");
+        send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+        return -1;
+    }
+
+    if (strlen(request) < 10 || strncmp(request, "GET /?id=", 9) != 0) {
```

Comment:
I think the original author did this to prevent a received request, that is shorter than the expected string to compare beyond what was received.

eg. request is of size 3, then you should not compare a length of 9.

Reply:
But now that both strings are properly terminated I can remove it. Otherwise it should have compared the read size instead of the string length.